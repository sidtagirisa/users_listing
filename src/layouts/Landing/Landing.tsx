import React, { useContext, useEffect } from 'react';
import { ActionType } from '../../constants';
import { StateContext } from '../../context/StateProvider';
import { getUsers } from '../../services/users';

const init = async (dispatch: React.Dispatch<any>): Promise<void> => {
  const users = await getUsers();
  dispatch({
    type: ActionType.UPDATE_USERS,
    data: users,
  });
  dispatch({
    type: ActionType.UPDATE_LOADING,
    data: false,
  });
};

export default function Landing(props: { children: JSX.Element }): JSX.Element {
  // @ts-ignore
  const { state, dispatch } = useContext(StateContext);
  const { loading } = state;
  const { children } = props;

  useEffect(() => {
    init(dispatch);
  }, [dispatch]);

  if (loading) {
    return (
      <React.Fragment>
        <div>loading...</div>
      </React.Fragment>
    );
  }
  return <React.Fragment>{children}</React.Fragment>;
}
