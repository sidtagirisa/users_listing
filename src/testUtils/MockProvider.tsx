import React from 'react';
import StateProvider from '../context/StateProvider/index';
import initialState from '../context/initialState';

export default function MockProvider(props: {
  additionalStateRequired?: object;
  children: JSX.Element;
}): JSX.Element {
  const { additionalStateRequired, children } = props;

  return (
    <StateProvider
      initialState={{
        ...initialState,
        ...additionalStateRequired,
      }}
    >
      {children}
    </StateProvider>
  );
}

MockProvider.defaultProps = {
  value: {
    state: {},
    dispatch: jest.fn(),
  },
};
