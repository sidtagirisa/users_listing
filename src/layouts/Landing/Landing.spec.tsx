import { act, render } from '@testing-library/react';
import request from 'axios';
import React from 'react';
import Landing from '.';
import { users } from '../../testUtils/fixtures';
import MockProvider from '../../testUtils/MockProvider';

describe('# Landing', () => {
  let requestMock: any;
  beforeEach(() => {
    requestMock = jest.spyOn(request, 'get');
  });

  it('should render basic component', async () => {
    requestMock.mockResolvedValue({ data: users });
    let container: any;

    await act(async () => {
      container = render(
        <MockProvider additionalStateRequired={{ loading: false }}>
          <Landing>
            <p>children</p>
          </Landing>
        </MockProvider>
      ).container;
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
