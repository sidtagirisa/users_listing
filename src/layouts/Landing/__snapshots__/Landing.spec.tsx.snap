// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`# Landing should render basic component 1`] = `
<p>
  children
</p>
`;
